subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/MavenResolver.java

```diff
@@ -220,8 +221,8 @@ public static MutableMavenModuleResolveMetadata processMetaData(MutableMavenModu
 
     private class MavenLocalRepositoryAccess extends LocalRepositoryAccess {
         @Override
-        protected void resolveModuleArtifacts(MavenModuleResolveMetadata module, BuildableComponentArtifactsResolveResult result) {
-            if (!module.getVariants().isEmpty()) {
+        protected void resolveModuleArtifacts(MavenModuleResolveMetadata module, ConfigurationMetadata variant, BuildableComponentArtifactsResolveResult result) {
+            if (!variant.requiresMavenArtifactDiscovery()) {
```

Comment:
There is a decision to be made here in the `MavenResolver` which artifact discovery strategy we use later when we actually get the artifacts.
The strategy we use for GMM and ivy is the `FixedComponentArtifacts` strategy. There we know exactly which artifacts we want and their location.

Now, artifacts (files) added by rules should be treated the same way as artifacts from GMM. So they should also use the `FixedComponentArtifacts` strategy.

Artifacts from pom metadata use a different strategy - `MetadataSourcedComponentArtifacts` - so they can discover artifacts not directly referenced in the pom (e.g. `module.jar` vs. `module.extension-from-packaging`).

Now, if we use a rule to modify a variant derived from POM, we have to make a decision which strategy to use as we suddenly have a mix of artifact types. Here we decide to use the `FixedComponentArtifacts` because that works with the file-based artifacts we define in the rule. It also works for pom artifacts, if they are the default (jar) - which also motivated the special `VariantFilesRules.isFilePathUnambiguous()` check.

Without the change here, we could not make a different decision for each variant. E.g., if I add a new variant for a _classifier_ artifact to a pom-based module, that would change the strategy for the existing `api` and `runtime` variants as well. Which might lead to not discovering their artifacts anymore without also adding a rule for them. Before, there was just no need to select different strategies for different variants, as all variants of a component would have the same types of artifacts.

I think it is ok to pass the `ConfigurationMetadta` (variant) down here together with the `ComponentMetadata` as it is now needed for the decision making.

_Note: that these are all really corner cases where things are published to maven with non-jar artifacts indicating this via a different packaging_.

Selecting an artifact from a different variant is not influenced by this code path. I think because we are going for the artifact directly, bypassing decisions made based on the selected component and variant.